File: mg01/game/tech/Assets/Script/Org/Ref/AutoRelease.cs
/* ==============================================================================
 * AutoReleasePool
 * @author jr.zeng
 * 2016/12/9 10:28:31
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace mg.org
{

    public class AutoRelease : BaseObject
    {

        List<Ref> m_autoList = new List<Ref>();

        HashSet<Ref> m_refHash = new HashSet<Ref>();  //所以ref的总表, add时记录, 要到dispose时才remove, auto release时不移除

        public AutoRelease()
        {

        }


        //-------∽-★-∽------∽-★-∽--------∽-★-∽数据管理∽-★-∽--------∽-★-∽------∽-★-∽--------//

        public void Add(Ref ref_)
        {
            if (m_refHash.Contains(ref_))
                return;
            m_refHash.Add(ref_);

            ref_.Retain(null);
            m_autoList.Add(ref_);
        }

        public void Remove(Ref ref_)
        {
            if (!m_refHash.Contains(ref_))
                return;
            m_refHash.Remove(ref_);
           
            if (m_autoList.Remove(ref_))
            {   //还在列表里面
                ref_.Release(null);
                Log.Assert("怎么可能还在AutoList里?");
            }
        }


        public void Excute()
        {
            if (m_autoList.Count == 0)
                return;

            Ref ref_;
            for(int i = m_autoList.Count-1;i>=0;--i)
            {
                ref_ = m_autoList[i];
                m_autoList.RemoveAt(i); //先移除再release,不然在remove里又会release一次
                ref_.Release(null);     //全部引用减1
            }

            //m_autoList.Clear();
        }


        //-------∽-★-∽------∽-★-∽--------∽-★-∽数据操作∽-★-∽--------∽-★-∽------∽-★-∽--------//

        //todo
        public void Dump()
        {

        }

    }


}

File: mg01/game/tech/Assets/Resources/LuaScript/src/org/ref/AutoRelease.lua
--AutoRelease
--@author jr.zeng
--2017年10月11日 上午10:28:12
local modname = "AutoRelease"
--==================global reference=======================

--===================namespace========================
local ns = "org"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using, modname)
--===================module property========================
local super = nil	--父类
_ENV[modname] = class(modname, super)
local AutoRelease = _ENV[modname]

--===================module content========================

function AutoRelease:__ctor()

    self.m_refDic = {}  --用来保留所有引用, 方便统计
    self.m_autoArr = new(Array) --自动释放列表
        
end

function AutoRelease:add(ref_)
    
    ref_:retain(false)
    self.m_autoArr:add(ref_)
    self.m_refDic[ref_] = ref_
	--nslog.print_t("添加到autoRelaease", ref_.referId)
end

function AutoRelease:remove(ref_)
    
    if not self.m_refDic[ref_] then
        return end
    self.m_refDic[ref_] = nil
    
    if self.m_autoArr:remove(ref_) then
        ref_:release(false)
        nslog.error("怎么可能还在autoArr里?")
    end
    
    nslog.debug("从autoRelease移除")
end

--执行自动释放
function AutoRelease:excute()
    
    if #self.m_autoArr == 0 then
        return end
    
    local ref
    local len = #self.m_autoArr
    for i=len, 1, -1 do
        ref = self.m_autoArr[i]
        self.m_autoArr:remove_at(i)  --先移除再release,不然在remove里又会release一次

	    --if ref:ref_cnt() <= 1 then
		 --   nslog.print_t('被AutoRelease了', ref.referId)
	    --end
	    --nslog.print_t("自动释放", ref.referId)

        ref:release(false)
    end
    self.m_autoArr:clear()
end

function AutoRelease:clear()
    
    self:excute()
    self.m_refDic = {}
end


function AutoRelease:dump()
    --TODO
end

return AutoRelease